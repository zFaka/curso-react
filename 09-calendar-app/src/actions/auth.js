import Swal from "sweetalert2";
import {fetchWithToken, fetchWithuotToken} from "../helpers/fetch";
import {types} from "../types/types";
import {eventLogout} from "./events";



export const startLogin = (email, password) => {

  return async(dispatch) => {

    const rsp = await fetchWithuotToken('auth', {email, password}, 'POST');
    const body = await rsp.json();

    if(body.ok){
      localStorage.setItem('token', body.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(login({
        uid:body.uid, 
        name:body.name
      }))
    } else {
      Swal.fire('Error', body.msg, 'error')
    }
  }
};



export const startRegister = (name, email, password ) => {
  return async(dispatch) => {

    const rsp = await fetchWithuotToken('auth/new', {name, email, password}, 'POST');
    const body = await rsp.json();

    if(body.ok){
      localStorage.setItem('token', body.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(login({
        uid:body.uid, 
        name:body.name
      }))
    } else {
      Swal.fire('Error', body.msg, 'error')
    }
  }
};



export const startChecking = () => {
  return async(dispatch) => {


    const rsp = await fetchWithToken('auth/renew');
    const body = await rsp.json();

    if(body.ok){
      localStorage.setItem('token', body.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(login({
        uid:body.uid, 
        name:body.name
      }))
    } else {
      dispatch(checkingFinish())
    }
  }
};



const checkingFinish = (user) => (
  {type:types.authLogin, 
    payload:user})



const login = (user) => ({

  type:types.authLogin, 
  payload:user
})



export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear()
    dispatch(eventLogout())
    dispatch(logout())
  }
};


const logout = () => ({type:types.authLogout})
