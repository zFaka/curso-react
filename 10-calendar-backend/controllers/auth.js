const {validationResult} = require("express-validator");
const User = require("../models/User");
const bcrypt = require('bcryptjs');
const {generateJWT} = require("../helpers/jwt");



const createUser = async(req, res) => {


    const {email, password} = req.body;



    try{

        let user = await User.findOne({email})



        if(user){
            return res.status(400).json({
                ok:false, 
                msg:'Exist a user with this email'
            })
        }


        user = new User(req.body);


        // Encrypt Pass
        const salt = bcrypt.genSaltSync();
        user.password = bcrypt.hashSync(password, salt)


        await user.save()


        // Generate JWT
        const token = await generateJWT(user.id, user.name);


        res.status(201).json({
            ok: true, 
            uid: user.id, 
            name: user.name, 
            token
        })


    } catch (err) {

        console.log(err) 

        res.status(500).json({
            ok:false, 
            msg:'Please contact with the admin'
        })
    }
}



const loginUser = async(req, res) => {


    const {email, password} = req.body;



    try{

        const user = await User.findOne({email})



        if(!user){
            return res.status(400).json({
                ok:false, 
                msg:"Doesn't exist a user with this email"
            })
        }

        //Validate Pass
        const validPass = bcrypt.compareSync(password, user.password);

        if (!validPass) {
            return res.status(400).json({
                ok:false, 
                msg:'This is not the password'})}


        // Generate JWT
        const token = await generateJWT(user.id, user.name);


        res.json({
            ok:true, 
            uid:user.id, 
            name:user.name, 
            token
        })


    } catch (err) {

        console.log(err) 

        res.status(500).json({
            ok:false, 
            msg:'Please contact with the admin'
        })
    }
}



const revalidateToken = async(req, res) => {

    const {uid, name} = req;
    const token = await generateJWT(uid, name);

    res.json({
        ok: true, 
        token
    })
};


module.exports = {
    createUser, 
    loginUser, 
    revalidateToken, 
}
