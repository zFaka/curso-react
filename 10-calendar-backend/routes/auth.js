/*
    User Router /Auth
    host+ /api/auth

*/


const {Router} = require('express');
const {check} = require('express-validator');
const {createUser, loginUser, revalidateToken} = require('../controllers/auth');
const {fieldValidator} = require('../middlewares/field-validator');
const {jwtValildator} = require('../middlewares/jwt-validator');
const router = Router();


router.post('/new', [ 
    check('name', 'Name is necessary').not().isEmpty().isLength({min:3}),
    check('email', 'Email is necessary').isEmail(), 
    check('password', 'Pass is necessary').isLength({min:6}), 
    fieldValidator
], 
    createUser)


router.post('/',[
    check('email', 'Email is necessary').isEmail(), 
    check('password', 'Pass is necessary').isLength({min:6}), 
    fieldValidator
], 
    loginUser)

router.get('/renew', 
    jwtValildator, 
    revalidateToken)

module.exports = router
