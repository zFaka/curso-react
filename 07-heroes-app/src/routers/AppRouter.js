import React, {useContext} from 'react';
import {
  BrowserRouter as Router, 
  Switch, 
} from 'react-router-dom';
import LoginScreen from '../components/login/LoginScreen';
import DashboardRouters from './DashboardRouters';
import PrivateRoute from './PrivateRoute';
import {AuthContext} from '../auth/AuthContext';
import PublicRoute from './PublicRoute';

const AppRouter = () => {


  const {user} = useContext(AuthContext);



  return (
    <Router>
      <div>
        <Switch>


          <PublicRoute
            exact path="/login"
            component={LoginScreen}
            isAuthenticated={user.logged}
          />

          <PrivateRoute 
            path="/" 
            component={DashboardRouters}
            isAuthenticated={user.logged}
          />
        </Switch>
      </div>
    </Router>
  )
}

export default AppRouter;
