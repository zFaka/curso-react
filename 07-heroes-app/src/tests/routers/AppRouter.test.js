import {mount, shallow} from "enzyme";
import {AuthContext} from "../../auth/AuthContext";
import AppRouter from "../../routers/AppRouter";

describe('Test AppRouter', () => {


  const contextValue = {
    dispatch: jest.fn(), 
    user: {logged:false}};


  test('Should show login if is not authenticated', () => {


    // value recive an Object with user and dispatch value
    // 1 value={Object}   recive an Object
    // 2 {user, dispatch} its an Object
    // 3 value={contextValue} in this case contextValue is an Object 
    // 4 value={{user, dispatch}} put Object inside value


    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter/>
      </AuthContext.Provider>
    );



    expect(wrapper).toMatchSnapshot()
  })



  test('Should show marvel if is authenticated', () => {


    const contextValue = {
      dispatch: jest.fn(), 
      user: {
        logged:true, 
        name:'faka'}};

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter/>
      </AuthContext.Provider>);



    // .nameOfClass is used for find class in components
    expect(wrapper.find('.navbar').exists())
      .toBe(true)
  })
})
