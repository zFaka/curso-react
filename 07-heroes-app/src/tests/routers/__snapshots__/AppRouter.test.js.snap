// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AppRouter Should show login if is not authenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      SignIn
    </h1>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      login
    </button>
  </div>
</div>
`;
